### DISCUSSION TITLE
[Question] How to draw image of substructures in the center of drawings?

### DISCUSSION THREAD

**Discussion by mohammad-saber** [rdkit/rdkit#4455] [Q&A]
Hi, thank you for reading my question. 

I am using the following code to draw two sub-structures of a parent molecule. The function `set_submol_position` finds the position of atoms in parent molecule, and set the same position for sub_mols.

```
from rdkit import Chem
from rdkit.Chem.Draw import rdMolDraw2D
from IPython.display import SVG
from rdkit.Chem import rdDepictor


def set_submol_position(mol_parent, mol):
    """
    This function finds the position of atoms in parent molecule, 
    and set the same position for sub_mols:       
        Compute 2D coordinates for the parent molecule and sub molecule
        Match substructure and copy the coordinates over
    """ 
    rdDepictor.Compute2DCoords(mol_parent)
    
    # Get the coordinates of substructure to built a conformer
    rdDepictor.Compute2DCoords(mol)

    # Get the index of the matched atoms
    matched_atoms_indexes = mol_parent.GetSubstructMatch(mol)

    # If there are more than one match, "matched_atoms_indexes" will be a tuple of tuples.
    # We check the first entry, if it is a tuple, we use the 1st entry.
    # If there is only one match, "matched_atoms_indexes" is a tuple and we use it as it is.
    if type(matched_atoms_indexes[0]) == tuple:   
        matched_atoms_index = matched_atoms_indexes[0]
    else: 
        matched_atoms_index = matched_atoms_indexes
    
    # Get positions of the matched atoms
    conf1 = mol_parent.GetConformer()
    pos = [list(conf1.GetAtomPosition(x)) for x in matched_atoms_index]

    # Set the original positions of parent molecule to substructure
    conf2 = mol.GetConformer()
    for n in range(len(matched_atoms_index)):
        conf2.SetAtomPosition(n, pos[n])


smiles_parent = 'c1ccc2ocnc2c1'
mol_parent = Chem.MolFromSmiles(smiles_parent)

sub_structures = ['c(c)c', 'o(c)c']
legends = sub_structures

# Read sub_structures
sub_mols = []
for smiles in sub_structures:
    
    if any(True for c in smiles if c.islower()):   # If there is any aromatic atom
        mol = Chem.MolFromSmiles(smiles, sanitize=False) 
        set_submol_position(mol_parent, mol)   # find the position of atoms in parent molecule, and set the same position for sub_mols           
        Chem.SanitizeMol(mol, sanitizeOps=(Chem.SanitizeFlags.SANITIZE_ALL^Chem.SanitizeFlags.SANITIZE_KEKULIZE^Chem.SanitizeFlags.SANITIZE_SETAROMATICITY))
        sub_mols.append(mol)

    else:
        mol = Chem.MolFromSmiles(smiles)  
        sub_mols.append(mol)


# Drawing
fullSize = (400, 200)   # size of image
subImgSize = (200, 200)
rdDepictor.Compute2DCoords(mol_parent)   

d2d = rdMolDraw2D.MolDraw2DSVG(fullSize[0], fullSize[1], subImgSize[0], subImgSize[1])
d2d.drawOptions().prepareMolsBeforeDrawing=False   # not to kekulize the bonds before drawing
d2d.DrawMolecules(sub_mols, legends=legends)   
d2d.FinishDrawing()
image = SVG(d2d.GetDrawingText())   
```

The result I get:

<img width="408" alt="1" src="https://user-images.githubusercontent.com/38423900/131239922-62072646-fde6-4363-bc8f-006c4cbb480c.png">

The code works fine. However, I would like to know how to draw sub-structures in the center of drawings, like sample image below:

![2](https://user-images.githubusercontent.com/38423900/131239928-0c0fac21-7863-4946-9d96-c35329927e0c.png)

**Comment by greglandrum** [comment]
@DavidACosgrove Do you think DrawMolecules() should center the molecules even if `prepareMolsBeforeDrawing` is False? I guess I can see arguments either way.

**Comment by DavidACosgrove** [comment]
Hi @Greg Landrum ***@***.***>,
It feels as though 99.9% of the time that would be a sensible thing to do.
I can imagine that from time to time it might be intensely frustrating.
You might want to put a picture in the corner of an image to leave room for
text added by some other means, for example.  How would it play with the
reaction drawing?  As I recall, those are built up from several component
drawings.  I guess an extra option to turn off automatic centring would be
more bloat in the drawing options for a minor gain.  It's relatively
straightforward to centre the drawing coordinates before passing them to
the drawing code.  Could we expose the internal function for that? Or maybe
it could be added as some Cookbook code for people to employ should the
need arise.

On Tue, Aug 31, 2021 at 5:35 PM Greg Landrum ***@***.***>
wrote:

> @DavidACosgrove <https://github.com/DavidACosgrove> Do you think
> DrawMolecules() should center the molecules even if
> prepareMolsBeforeDrawing is False? I guess I can see arguments either way.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/rdkit/rdkit/discussions/4455#discussioncomment-1260965>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACGF2FWKSSGIQLOPBQUZ26DT7UAGTANCNFSM5C734LFA>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
>


-- 
David Cosgrove
Freelance computational chemistry and chemoinformatics developer
http://cozchemix.co.uk
